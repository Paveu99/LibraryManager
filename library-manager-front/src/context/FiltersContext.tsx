import { createContext } from "react";

export interface FiltersContextType {
    chosenFilters: { year: string, month: string } | null;
    setChosenFilters: (payload: { year: string, month: string } | null) => void;
}

export const FiltersContext = createContext<FiltersContextType | null>(null);
