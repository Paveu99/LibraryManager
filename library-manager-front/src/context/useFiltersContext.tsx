import { useContext } from "react";
import { FiltersContext, } from "./FiltersContext";

export const useFiltersContext = () => {
    const context = useContext(FiltersContext);

    if (!context) {
        throw new Error("useCategoryContext must be used within a CategoryContext.Provider");
    }

    return context;
};
