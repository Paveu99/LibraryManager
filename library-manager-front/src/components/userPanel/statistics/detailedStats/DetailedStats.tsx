import { FormControl, InputLabel, MenuItem, Select } from "@mui/material";
import { useEffect, useState } from "react";
import { useGetPreciseUserStatsQuery } from "../../../../queries/stats/useGetPreciseUserStatsQuery";
import { axisClasses, BarChart } from "@mui/x-charts";
import { useFiltersContext } from "../../../../context/useFiltersContext";

type Props = {
    allTime: {
        month?: string;
        year?: number;
        rented: number;
        overdue: number;
        returned: number;
    }[]
}

export const DetailedStats = ({ allTime }: Props) => {
    const [year, setYear] = useState<string>('');
    const [month, setMonth] = useState<string>('');
    const { data, isLoading } = useGetPreciseUserStatsQuery<{ month: string; rented: number; overdue: number; returned: number }[]>(Number(year));
    const { setChosenFilters } = useFiltersContext();

    const handleChange = (value: string) => {
        setYear(value);
        setMonth('');
    };

    const handleMonthChange = (value: string) => {
        setMonth(value);
    };

    const chartSetting = {
        yAxis: [
            {
                label: 'Amount of books',
            },
        ],
        width: 600,
        height: 200,
        sx: {
            [`.${axisClasses.left} .${axisClasses.label}`]: {
                transform: 'translate(-5px, 0)',
            },
        },
    };

    useEffect(() => {
        setChosenFilters({
            year: year,
            month: month
        })
    }, [year, month])

    const filteredData = data?.filter(item => month ? item.month === month : true);

    if (isLoading) return <p>Loading detailed stats...</p>;

    return (
        <div>
            <FormControl sx={{ m: 1, minWidth: 120 }} size="small">
                <InputLabel id="demo-select-small-label">Year</InputLabel>
                <Select
                    labelId="demo-select-small-label"
                    id="demo-select-small"
                    value={year}
                    label="Year"
                    onChange={(e) => handleChange(e.target.value)}
                >
                    <MenuItem value="">
                        <em>None</em>
                    </MenuItem>
                    {allTime.map((el, index) => (<MenuItem value={el.year} key={index}>{el.year}</MenuItem>))}
                </Select>
            </FormControl>
            {year && <FormControl sx={{ m: 1, minWidth: 120 }} size="small">
                <InputLabel id="demo-select-small-label">Month</InputLabel>
                <Select
                    labelId="demo-select-small-label"
                    id="demo-select-small"
                    value={month}
                    label="Month"
                    onChange={(e) => handleMonthChange(e.target.value)}
                >
                    <MenuItem value="">
                        <em>None</em>
                    </MenuItem>
                    {data?.map((el, index) => (<MenuItem value={el.month} key={index}>{el.month}</MenuItem>))}
                </Select>
            </FormControl>}
            {
                !year ? <BarChart
                    dataset={allTime.map(el => ({ ...el, year: el.year || 'Unknown Year' }))}
                    xAxis={[{ scaleType: 'band', dataKey: 'year' }]}
                    series={[
                        { dataKey: 'rented', label: 'Rented' },
                        { dataKey: 'returned', label: 'Returned' },
                        { dataKey: 'overdue', label: 'Overdue' }
                    ]}
                    {...chartSetting}
                /> :
                    <BarChart
                        dataset={filteredData?.map(el => ({
                            ...el,
                            month: el.month || 'Unknown Month',
                        }))}
                        xAxis={[{ scaleType: 'band', dataKey: 'month' }]}
                        series={[
                            { dataKey: 'rented', label: 'Rented' },
                            { dataKey: 'returned', label: 'Returned' },
                            { dataKey: 'overdue', label: 'Overdue' }
                        ]}
                        {...chartSetting}
                    />
            }
        </div>
    );
};
