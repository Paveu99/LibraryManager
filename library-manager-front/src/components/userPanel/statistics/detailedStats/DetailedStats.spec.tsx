import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render, screen } from "@testing-library/react";
import { DetailedStats } from "./DetailedStats";
import { FiltersContext } from "../../../../context/FiltersContext";
import { useState } from "react";
import { vi } from "vitest";
import userEvent from "@testing-library/user-event";

vi.mock("../../../../queries/stats/useGetPreciseUserStatsQuery", () => ({
    useGetPreciseUserStatsQuery: vi.fn(() => ({
        data: object1.allTime,
        isLoading: false,
        error: null,
    })),
}));

const object1 = {
    allTime: [
        { year: 2020, rented: 0, overdue: 0, returned: 1 },
        { year: 2021, rented: 0, overdue: 0, returned: 4 },
        { year: 2022, rented: 0, overdue: 2, returned: 3 },
        { year: 2023, rented: 0, overdue: 0, returned: 2 },
        { year: 2024, rented: 0, overdue: 1, returned: 3 },
        { year: 2025, rented: 3, overdue: 3, returned: 9 }
    ]
};

const TestWrapper = ({ children }: { children: React.ReactNode }) => {
    const queryClient = new QueryClient();
    const [chosenFilters, setChosenFilters] = useState<{ year: string, month: string } | null>(null);

    return (
        <QueryClientProvider client={queryClient}>
            <FiltersContext.Provider value={{ chosenFilters, setChosenFilters }}>
                {children}
            </FiltersContext.Provider>
        </QueryClientProvider>
    );
};

const renderWithProviders = (ui: React.ReactNode) => render(<TestWrapper>{ui}</TestWrapper>);

const selectOption = async (label: string, optionText: string) => {
    const selectElement = screen.getByLabelText(label);
    await userEvent.click(selectElement);

    const optionElements = await screen.findAllByText(optionText);
    const option = optionElements.find((el) => el.tagName.toLowerCase() === "li");

    expect(option).toBeDefined();
    if (option) {
        await userEvent.click(option);
    }
};

describe("DetailedStats", () => {
    it("renders component with correct all-time data", async () => {
        renderWithProviders(<DetailedStats allTime={object1.allTime} />);
        await screen.findByText("2020");
    });

    it("renders correctly when there is no data", () => {
        renderWithProviders(<DetailedStats allTime={[]} />);
        expect(screen.getByText("No data to display")).toBeInTheDocument();
    });

    it("allows user to select a year and updates displayed data", async () => {
        renderWithProviders(<DetailedStats allTime={object1.allTime} />);
        await selectOption("Year", "2025");
    });

    it("allows user to select a year and a month, then updates displayed data", async () => {
        renderWithProviders(<DetailedStats allTime={object1.allTime} />);
        await selectOption("Year", "2025");
        await selectOption("Month", "January");
    });
});
