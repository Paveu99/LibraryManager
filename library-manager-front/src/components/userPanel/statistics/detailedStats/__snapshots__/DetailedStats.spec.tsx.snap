// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DetailedStats > matches snapshot 1`] = `
<DocumentFragment>
  <p
    class="warnings"
  >
    Loading...
  </p>
</DocumentFragment>
`;
